src/net_processing.cpp

```diff
@@ -2470,9 +2485,11 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
     if (strCommand == NetMsgType::TX) {
         // Stop processing the transaction early if
         // We are in blocks only mode and peer is either not whitelisted or whitelistrelay is off
-        if (!g_relay_txes && !pfrom->HasPermission(PF_RELAY))
+        // or if this peer is supposed to be a block-relay-only peer
+        if ((!g_relay_txes && !pfrom->HasPermission(PF_RELAY)) || (pfrom->m_tx_relay == nullptr))
```

Comment:
This doesn't match the `fBlocksOnly` test for `::INV` -- if it did it would be:

```
((!g_relay_txes || pfrom->m_tx_relay == nullptr) && !pfrom->HasPermission(PF_RELAY))
```

I think the logic here is probably better, though maybe there should be a check to prevent choosing a node for blocks-relay if it would have PF_RELAY perms?

Reply:
I believe it's impossible for an outbound peer to be whitelisted and have PF_RELAY permissions, so these are the same...

Reply:
I think you're right; `HasPermission()` checks `m_permissionFlags` and the only way m_permissionFlags is anything but PF_NONE is via `AcceptConnection()`, so inbound only.